port/unix/omrmmap.c

```diff
@@ -50,11 +49,98 @@
 #include "protect_helpers.h"
 #include "omrportpriv.h"
 
+#if defined(J9ZOS390)
+#include "omrvmem.h"
+#if !defined(MAP_FAILED)
+#define MAP_FAILED ((void*)-1)
+#endif /* !defined(MAP_FAILED) */
+#endif /* defined(J9ZOS390) */
+
 #if defined(AIXPPC)
 #include <sys/shm.h>
 #include <sys/vminfo.h>
 #endif/*AIXPPC*/
 
+#if defined(J9ZOS390)
+static J9MmapHandle *
+omrmmap_read_map_file(struct OMRPortLibrary *portLibrary, intptr_t file, uint64_t offset, uintptr_t size, uint32_t category)
+{
+	/* default implementation will allocate memory and read the file into it */
+	uintptr_t numBytesRead = 0;
+	intptr_t rc = 0;
+	void *mappedMemory = NULL;
+	void *allocPointer = NULL;
+	J9MmapHandle *returnVal = NULL;
+
+	Trc_PRT_mmap_map_file_default_entered(file, size);
+
+	if (file == -1) {
+		Trc_PRT_mmap_map_file_default_badfile();
+		return NULL;
+	}
+
+	if (0 == size) {
+		struct stat buf;
+
+		memset(&buf, 0, sizeof(buf));
+		if (-1 == fstat(file - FD_BIAS, &buf)) {
+			Trc_PRT_mmap_map_file_unix_filestatfailed();
+			portLibrary->error_set_last_error(portLibrary, errno, OMRPORT_ERROR_MMAP_MAP_FILE_STATFAILED);
+			return NULL;
+		}
+		size = buf.st_size;
+	}
+
+	if (0 != portLibrary->file_seek(portLibrary, file, offset, EsSeekSet)) {
+		Trc_PRT_mmap_map_seek_failed(offset);
+		return NULL;
+	}
+
+	/* ensure that allocated memory is 8 byte aligned, just in case it matters */
+	allocPointer = portLibrary->mem_allocate_memory(portLibrary, size + 8, OMR_GET_CALLSITE(), category);
+	Trc_PRT_mmap_map_file_default_allocPointer(allocPointer, size + 8);
```

Comment:
Where does 8 come from here? Is it `alignof(double)` or something else?

Reply:
Unfortunately I don't know the history of where 8 is from. The code is just refactored from the zos390/omrmmap.c.

Reply:
It must be to align 64-bit values, although not specifically `double`.

Reply:
Yes, I think using `ROUND_UP_TO_POWEROF2()` would make it clearer, but I'd like also to see a comment explaining why we're rounding up the both size and the pointer to that power of 2 (i.e. an explanation for that specific power).

Reply:
Updated.

Reply:
@keithc-ca, do you have more comment on this ?